Added more escaping characters for HTML

Ryuuzakis opened the pull request:
Hi.

I've parsed a list of html escape characters (that you can find [here](http://www.theukwebdesigncompany.com/articles/entity-escape-characters.php)), and added them to the html.cr file.

Hope it's good :smile:

marceloboeira commented:
@Ryuuzakis 

Tests are failing, take a look: https://travis-ci.org/manastech/crystal/jobs/110107440
```
crystal spec ./spec/std/html_spec.cr:6 # HTML .escape does not change a safe string
crystal spec ./spec/std/html_spec.cr:12 # HTML .escape escapes dangerous characters from a string
```

asterite commented:
I'm not sure we need to escape those characters. The current implementation contains problematic characters that you always want to escape in HTML, like `<`, `>`, `'`, `"` and `&`. I think I needed to escape other chars, like accented spanish chars, when I was programming in PHP maybe 8 years ago, but I think now that's not needed anymore.

Ryuuzakis commented:
Oh right ! I forgot to launch the tests, my bad.

So you say we should leave it as it was ? Or maybe remove some unneeded escaped characters ? I just lazily let them all, but I could sort them a bit to make them pass the tests.

asterite commented:
I think the current list is fine. But I honestly don't know when and what characters need to always be escaped in HTML.

Ryuuzakis commented:
Okay, so I've searched a bit, and it seems that we need a little more characters escaping that the set there already is.

According to [this blog article](http://wonko.com/post/html-escaping), which refers and explains content from the [OWASP](https://www.owasp.org/index.php/XSS_%28Cross_Site_Scripting%29_Prevention_Cheat_Sheet#XSS_Prevention_Rules), you need more characters to prevent XSS injections.

I will update my branch by limiting the characters to what is really needed when I have some time to spare :)

luislavena commented on src/html.cr:
```diff
@@ -1,11 +1,372 @@
-    '<'  => "&lt;",
-    '>'  => "&gt;",
-  }
+SUBSTITUTIONS = {
```
Looks like indentation miss? `SUBSTITUTIONS` should be nested (by 2 spaces) inside `HTML` and all the elements in the hash also nested (by 2 spaces) from `SUBSTITUTIONS` definition.

You can use Crystal's `format` to correct that (`crystal tool format src/html.cr`)

Ryuuzakis replied:
Thanks for the tip, I'll look into it !

luislavena commented:
Perhaps we should look at other implementations of HTML escaping before decide? Basically the substitution list looks too big now :stuck_out_tongue_winking_eye:

ysbaddaden commented:
@Ryuuzakis goot catch. We should indeed escape some more characters to be (almost) safe.

Ryuuzakis commented:
I've done the modifications I talked about earlier, and added a new test case for the escape method.
Telle me what you think about it :)

asterite commented:
@Ryuuzakis Thanks!

Do we need to convert spaces to `&nbsp;`, though? I also don't understand why the spaces are not converted in your spec...

Ryuuzakis commented:
The space that is protected is the non-breaking space, this is why the classical space is not converted.

asterite commented:
@Ryuuzakis I see. I think it's very confusing if we leave it like that. Another confusing case is the single quote, it looks like `'''`, which looks like a typo. Better change `'''` to `'\''` and the non-breaking space to ``\u{a0}'` or `160.chr`.

Ryuuzakis commented:
@asterite I agree with you, your solution was better, so I just implemented it. I've also added a test case to test the non-breaking spaces and normal spaces.

luislavena commented on src/html.cr:
```diff
@@ -1,10 +1,24 @@
+    '`'      => "&#96;",
+    '{'      => "&#123;",
+    '}'      => "&#125;",
+    '\u{a0}' => "&nbsp",
```
Perhaps missing `;` at the end?

asterite replied:
@luislavena Good catch

Ryuuzakis replied:
The `;` is not needed for it to work. Do you want me to add it, for consistency purpose ?

luislavena replied:
@Ryuuzakis afaik all HTML entities must be enclosed between `&` and `;`, at least that is what W3C says, if that works in some browsers that will be fun.

Either way, I think for consistency, yes, let's have them according to W3C.

Thank you.

Ryuuzakis replied:
@luislavena Guess you've got a valid point. It's done :)

asterite commented:
@Ryuuzakis Thank you! We are almost done here. The indentation is wrong in the spec file. Could you run `crystal tool format` to fix that?

Ryuuzakis commented:
Done !

ysbaddaden commented:
Thank you!
